### PR TITLE
Replace doAsync calls with Kotlin coroutines in SearchViewModel

### PR DESCRIPTION
**Pull request by afollestad**
Replace AsyncTask with coroutines, which are scoped to the view model so execution seizes when the view model is cleared (the screen it belongs to goes away).

Part 1 of https://github.com/naman14/TimberX/issues/6.

### CODE REVIEW

**Comment by afollestad** [comment]
CC @naman14 @Zhuinden

File: app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt [hunk 1]
```diff
@@ -24,40 +25,49 @@ import com.naman14.timberx.models.Song
...
+import kotlinx.coroutines.withContext
 
-class SearchViewModel(val context: Context) : ViewModel() {
+class SearchViewModel(val context: Context) : ScopedViewModel(Main) {
 
     private val searchData = SearchData()
     private val _searchLiveData = MutableLiveData<SearchData>()
 
-    val searchLiveData = _searchLiveData
+    val searchLiveData: LiveData<SearchData> = _searchLiveData
```
Comment (afollestad) [inline app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt]: This was formally leaking the full `MutableLiveData`

File: app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt [hunk 2]
```diff
@@ -24,40 +25,49 @@ import com.naman14.timberx.models.Song
...
 
     fun search(query: String) {
         if (query.length >= 3) {
-            doAsyncPostWithResult(handler = {
-                SongsRepository.searchSongs(context, query, 10)
-            }, postHandler = {
-                if (it!!.isNotEmpty())
-                    searchData.songs = ArrayList(it)
+            launch {
+                val songs = withContext(IO) {
```
Comment (afollestad) [inline app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt]: The fancy thing about coroutines is it's not just threading. Kotlin generates special code under the hood that handles context switching and line-by-line execution. And it uses thread pools which are FAR cheaper than actual threads.

File: app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt [hunk 3]
```diff
@@ -24,40 +25,49 @@ import com.naman14.timberx.models.Song
...
+            }
 
-            doAsyncPostWithResult(handler = {
-                AlbumRepository.getAlbums(context, query, 7)
-            }, postHandler = {
-                if (it!!.isNotEmpty())
-                    searchData.albums = ArrayList(it)
+            launch {
+                val albums = withContext(IO) {
+                    AlbumRepository.getAlbums(context, query, 7)
```
Comment (afollestad) [inline app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt]: This executes on the IO dispatcher

File: app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt [hunk 4]
```diff
@@ -24,40 +25,49 @@ import com.naman14.timberx.models.Song
...
+                }
+                if (albums.isNotEmpty()) {
```
Comment (afollestad) [inline app/src/main/java/com/naman14/timberx/ui/viewmodels/SearchViewModel.kt]: We return to the Main dispatcher here

File: app/src/main/java/com/naman14/timberx/ui/viewmodels/base/ScopedViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import kotlinx.coroutines.launch
+import org.jetbrains.annotations.TestOnly
+import kotlin.coroutines.CoroutineContext
+
+/** @author Aidan Follestad (@afollestad) */
+abstract class ScopedViewModel(
+    private val mainDispatcher: CoroutineDispatcher
+) : ViewModel() {
+
+    private val job = Job()
```
Comment (Zhuinden) [inline app/src/main/java/com/naman14/timberx/ui/viewmodels/base/ScopedViewModel.kt]: Do you by chance know if this has to be `Job` or `SupervisorJob`? I've seen both.

Comment (afollestad) [reply app/src/main/java/com/naman14/timberx/ui/viewmodels/base/ScopedViewModel.kt]: Job is what I use in my production code - and what we've come to use at @square :)

Comment (Zhuinden) [reply app/src/main/java/com/naman14/timberx/ui/viewmodels/base/ScopedViewModel.kt]: ok then i'll trust your words on this :slightly_smiling_face:

**Comment by naman14** [comment]
Thanks!, Let me read about coroutines in detail first, will be able to merge after that

**Comment by afollestad** [comment]
Since this was on my fork, I'm going to re-open with a branch in this repo

**Pull request closed without merging**